src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+	let meterReading1 = 0;
+	let meterReading2 = 0;
+
+	let startTimestamp = moment(0);
+	let endTimestamp = moment(0);
+
+	let currentReading = new Reading(meterID, -1, startTimestamp, endTimestamp);
+	let prevReading = currentReading;
+	let readingOK = true;
+	for (let index = readingRepetition; index <= rows.length; ++index) {
+		// To read data where same reading is repeated. Like E-mon D-mon meters
+		if ((index - readingRepetition) % readingRepetition === 0) {
+			if (onlyEndTime) {
+				// The startTimestamp of this reading is the endTimestamp of the previous reading
+				startTimestamp = prevReading.endTimestamp;
+				endTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			else {
+				startTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+				endTimestamp = moment(rows[index - readingRepetition][2], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			if (isCumulative && isFirst(prevReading.endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing with cumulative data.';
+			}
+			if (onlyEndTime && isFirst(prevReading.endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing only with endTimestamps.';
+			}
+			if (readingOK && startTimestamp.isSameOrAfter(endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The reading end time is not after the start time.';
+				if (onlyEndTime) {
+					errMsg += ' The start time came from the previous readings end time.';
+				}
+			}
+			if (readingOK) {
+				// Check that startTimestamp is not before the previous endTimestamp
+				if (onlyEndTime && endTimestamp.isSameOrBefore(prevReading.endTimestamp)) {
+					readingOK = false;
+					errMsg = 'The reading is not after the previous reading with only end time given so we must drop the reading.';
+				}
+				else if (startTimestamp.isBefore(prevReading.endTimestamp)) {
+					if (isCumulative) {
+						readingOK = false;
+						errMsg = 'The reading start time is before the previous endTime and cumulative so OED cannot use this reading.'
+					}
+					else if (!isFirst(prevReading.endTimestamp)) {
+						//Only treat this as a warning since the readings may be sent in a different order.
+						errMsg = 'The current reading startTime is not after the previous reading\'s end time. This is expected on the first reading ever.';
+					}
+				}
+			}
+			if (readingOK && Math.abs(startTimestamp.diff(prevReading.endTimestamp)) > Tgap) {
+				if (isCumulative) {
+					readingOK = false;
+					errMsg = 'The end of the previous reading is too far from the start of the next readings in cumulative data so drop the readings.';
+				}
+				else if (!isFirst(prevReading.endTimestamp)) {
+					//Only treat this as a warning since we expect a gap in the first ever reading.
+					errMsg = 'There is a gap in time between this reading and the previous reading.';
```

Comment:
I've gone ahead and done this. However, I have some doubts because it currently never calls this warning if the reading is the first reading which is the only case (I think) we expect an unexpected gap.

Therefore, I think this warning is only called in the case there is an actual unexpected gap in readings which I think is a valid error that we might want to consider doing something with.